phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java

```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
     @Override
     public synchronized Tuple next() throws SQLException {
         initScanner();
-        Tuple t = scanIterator.next();
-        return t;
+        try {
+            lastTuple = scanIterator.next();
+            if (lastTuple != null) {
+                ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+                lastTuple.getKey(ptr);
+            }
+        } catch (SQLException e) {
+            try {
+                throw ServerUtil.parseServerException(e);
+            } catch(StaleRegionBoundaryCacheException e1) {
+                if(scan.getAttribute(NON_AGGREGATE_QUERY)!=null) {
+                    Scan newScan = ScanUtil.newScan(scan);
+                    if(lastTuple != null) {
+                        lastTuple.getKey(ptr);
+                        byte[] startRowSuffix = ByteUtil.copyKeyBytesIfNecessary(ptr);
+                        if(ScanUtil.isLocalIndex(newScan)) {
+                            newScan.setAttribute(SCAN_START_ROW_SUFFIX, ByteUtil.nextKey(startRowSuffix));
+                        } else {
+                            newScan.setStartRow(ByteUtil.nextKey(startRowSuffix));
```

Comment:
See other comment below, but how about we serialize the row key we found which is out-of-region in the StaleRegionBoundaryCacheException instead of mucking with the startRowSuffix?